### PR TITLE
3D compute metadata assets

### PR DESCRIPTION
**Pull request by swheaton**
## What changes are proposed in this pull request?

- Add `SceneMetadata` which adds children asset sizes to `size_bytes` for fo3d files.
   - `asset_counts` field is a dict mapping extension to number of assets with that extension. For example, `{"obj": 8, "pcd": 1, "jpeg": 6}`
   - de-dupes asset paths within the same fo3d file
   - does NOT de-dupe across samples so sum of `size_bytes` for all samples may not be the same as the media size on export. A common case would be a background image or cube (6 images).
- Added FOU3D utility `get_scene_asset_paths()`, pulled in from a Teams PR by @brimoor 

both use `fos.run()` to run as tasks so we don't bring too much into memory at once. Protects against many giant .fo3d files.

## How is this patch tested? If it is not, please explain why.

-added tests for both functions

```python
# sorry don't have a public zoo option right now
ds = fo.load_dataset("my-3d-dataset")
ds.compute_metadata()
# see metadata is accurate. previously would just be size of .fo3d file

import fiftyone.utils.utils3d as fou3d
fou3d.get_scene_asset_paths(ds.values('filepath'))

# looks something like
{
  "/path/to/sample.fo3d": ["/path/to/absolute.pcd", "relative.obj"],
  "/path/to/sample2.fo3d": ...
}
```

### Is this a user-facing change that should be mentioned in the release notes?

-   [x] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

## Summary by CodeRabbit

- **New Features**
  - Introduced `SceneMetadata` for enhanced metadata management of 3D scenes.
  - Added a function to resolve file paths to their absolute forms, improving file handling across the platform.
  - Enhanced functionalities for extracting and managing asset paths from 3D scenes.

- **Tests**
  - Implemented new unit tests for metadata and 3D utilities to ensure robustness and reliability.

### CODE REVIEW

File: fiftyone/utils/utils3d.py [hunk 1]
```diff
@@ -738,6 +782,42 @@ def compute_orthographic_projection_image(
     return image, metadata
 
 
+def _get_pcd_filepath_from_scene(scene_path: str):
```
Comment (swheaton) [inline fiftyone/utils/utils3d.py:785]: did not really change just moved to match other PR

File: fiftyone/utils/utils3d.py [hunk 2]
```diff
@@ -439,35 +439,81 @@ class OrthographicProjectionMetadata(DynamicEmbeddedDocument, fol._HasMedia):
...
+    asset_paths = scene.get_asset_paths()
+
+    if abs_paths:
+        # Convert any relative-to-scene paths to absolute
+        scene_dir = os.path.dirname(original_scene_path)
+        for i, asset_path in enumerate(asset_paths):
+            if not fos.isabs(asset_path):
+                asset_paths[i] = fos.join(scene_dir, asset_path)
+
+    return asset_paths
```
[the remark this answers is not in the document]
Comment (swheaton) [reply fiftyone/utils/utils3d.py:472]: accepted in 2b4bf64fcaf1b15675fec4edc9a6786223d5dc3a

File: tests/unittests/metadata_tests.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+| `voxel51.com <https://voxel51.com/>`_
+|
+"""
+import json
+import os
+import tempfile
+import unittest
+
+import fiftyone.core.metadata as fom
+import fiftyone.core.storage as fos
```
[the remark this answers is not in the document]
Comment (swheaton) [reply tests/unittests/metadata_tests.py:14]: accepted in 6b226eaa9539782c07edaac0a80083b1963ed76c

File: fiftyone/core/metadata.py [hunk 4]
```diff
@@ -75,6 +78,70 @@ def _build_for_url(cls, url, mime_type=None):
         return cls(size_bytes=size_bytes, mime_type=mime_type)
 
 
+class SceneMetadata(Metadata):
+    """Class for storing metadata about 3D scene samples.
+
+    Args:
+        size_bytes (None): the size of scene definition and all children
+            assets on disk, in bytes
+        mime_type (None): the MIME type of the image
```
Comment (sashankaryal) [inline fiftyone/core/metadata.py:87]: nit: add more clarification on this arg?

File: fiftyone/core/metadata.py [hunk 5]
```diff
@@ -75,6 +78,70 @@ def _build_for_url(cls, url, mime_type=None):
...
+            if not fos.isabs(asset_path):
+                asset_path = fos.join(scene_dir, asset_path)
+            asset_paths[i] = fos.resolve(asset_path)
+            file_type = pathlib.Path(asset_path).suffix[1:]
+            asset_counts[file_type] += 1
+
+        # Dedupe asset paths within this single scene
+        asset_paths = list(set(asset_paths))
+        print(asset_paths)
+
```
Comment (sashankaryal) [inline fiftyone/core/metadata.py:126]: should we remove this print statement?

**Review by sashankaryal — APPROVED** [review]
👍

File: fiftyone/core/metadata.py (hunk 4 above)
Comment (swheaton) [reply fiftyone/core/metadata.py:87]: ah yes copy pasta. It is ignored and always set to `application/octet-stream` (didn't know what else to do)

File: fiftyone/core/metadata.py (hunk 5 above)
Comment (swheaton) [reply fiftyone/core/metadata.py:126]: dangit thanks

**Pull request merged**